File: Shortest_Path_Algo/1300-critical-connections-in-a-network/critical-connections-in-a-network.cpp
class Solution {
public:
    void dfs(int node, int parent, vector<vector<int>>& graph, vector<int>& disc, vector<int>& low, vector<vector<int>>& bridges, int& time) {
        disc[node] = low[node] = time++;
        for (int neighbor : graph[node]) {
            if (neighbor == parent) continue;
            if (disc[neighbor] == -1) {
                dfs(neighbor, node, graph, disc, low, bridges, time);
                low[node] = min(low[node], low[neighbor]);
                if (low[neighbor] > disc[node]) {
                    bridges.push_back({node, neighbor});
                }
            } else {
                low[node] = min(low[node], disc[neighbor]);
            }
        }
    }

    vector<vector<int>> criticalConnections(int n, vector<vector<int>>& connections) {
        vector<vector<int>> graph(n);
        vector<int> disc(n, -1), low(n, -1);
        vector<vector<int>> bridges;
        int time = 0;

        for (auto& conn : connections) {
            graph[conn[0]].push_back(conn[1]);
            graph[conn[1]].push_back(conn[0]);
        }

        for (int i = 0; i < n; i++) {
            if (disc[i] == -1) {
                dfs(i, -1, graph, disc, low, bridges, time);
            }
        }

        return bridges;
    }
};

File: Shortest_Path_Algo/1300-critical-connections-in-a-network/critical-connections-in-a-network.java
class Solution {
    private int time = 0;

    public void dfs(int node, int parent, List<List<Integer>> graph, List<List<Integer>> bridges, List<Integer> disc, List<Integer> low) {
        disc.set(node, time);
        low.set(node, time);
        time++;
        for (int neighbor : graph.get(node)) {
            if (neighbor == parent) {
                continue;
                }
            if (disc.get(neighbor) == -1) {
                dfs(neighbor, node, graph, bridges, disc, low);
                low.set(node, Math.min(low.get(node), low.get(neighbor)));
               if (low.get(neighbor) > disc.get(node)) {
                    bridges.add(Arrays.asList(node, neighbor));
                }
            } else {
               low.set(node, Math.min(low.get(node), disc.get(neighbor)));
            }
        }
    }

    public List<List<Integer>> criticalConnections(int n, List<List<Integer>> connections) {
        List<List<Integer>> graph = new ArrayList<>();
        for (int i = 0; i < n; i++) {
            graph.add(new ArrayList<>());
        }
        for (List<Integer> conn : connections) {
            int u = conn.get(0);
            int v = conn.get(1);
            graph.get(u).add(v);
            graph.get(v).add(u);
        }

        List<Integer> disc = new ArrayList<>(Collections.nCopies(n, -1));
        List<Integer> low = new ArrayList<>(Collections.nCopies(n, -1));
        List<List<Integer>> bridges = new ArrayList<>();

        for (int i = 0; i < n; i++) {
            if (disc.get(i) == -1) {
                dfs(i, -1, graph, bridges, disc, low);
            }
        }

        return bridges;
    }
}